Implement derivatives across intervals for aggregate queries

jsternberg opened the pull request:
For aggregate queries, derivatives will now alter the start time to one
interval behind and will use that interval to find the derivative of the
first point instead of giving no value for that interval.

This does not apply to raw queries yet.

Fixes #3247. Contributes to #5943.

jsternberg commented:
@benbjohnson @jwilder 

Also @nathanielc since this changes a public API that may be used by Kapacitor.

nathanielc commented:
@jsternberg The API change is OK with me.

benbjohnson commented:
@jsternberg Overall, lgtm. Can youc comment the `allowNil` in a little more detail? I think I understand it but I just want to see it documented for the future.

jsternberg commented:
I removed the `allowNil` section and modified derivative and difference to use the stream iterators that were used for moving average. I also got rid of the `allowNil` functionality and nil values, even if they are before the current interval, will just be discarded the same as if they were in the middle or the end.

This should now work for all aggregate queries.

sharang commented:
@jsternberg will you merge it into 0.12?

jsternberg commented:
@sharang unfortunately 0.12 has already been released and this wasn't ready in time to make feature cutoff. It's currently awaiting review and will likely go into 0.13 when we release that.

toddboom commented:
@jsternberg is anything else needed on this, or are you just waiting for some upvotes?

jsternberg commented:
Just waiting. This should work for aggregate queries which should alleviate at least some of the pain from this.

toddboom commented:
this seems good to me, though i'd like @benbjohnson to get some eyes on it since i'm not the most capable assessor of query engine changes

toddboom commented:
oh, ben already looked at it. i guess it's good to merge? :+1:

jsternberg commented:
It could use another look. @benbjohnson reviewed an older version of this.

benbjohnson commented:
👍
